File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperFactory.java
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding;

import java.util.EnumMap;
import java.util.Map;

import lombok.experimental.UtilityClass;
import no.nav.melosys.eessi.models.SedType;

@UtilityClass
public class MelosysEessiMeldingMapperFactory {

    private static final Map<SedType, MelosysEessiMeldingMapper> mappers = new EnumMap<>(SedType.class);
    private static final MelosysEessiMeldingMapper defaultMapper = new DefaultMapper();

    static {
        mappers.put(SedType.A001, new MelosysEessiMeldingMapperA001());
        mappers.put(SedType.A002, new MelosysEessiMeldingMapperA002());
        mappers.put(SedType.A003, new MelosysEessiMeldingMapperA003());
        mappers.put(SedType.A009, new MelosysEessiMeldingMapperA009());
        mappers.put(SedType.A010, new MelosysEessiMeldingMapperA010());
        mappers.put(SedType.A011, new MelosysEessiMeldingMapperA011());
        mappers.put(SedType.X006, new AdministrativSedMapperX006());
    }

    public static MelosysEessiMeldingMapper getMapper(SedType sedType) {
        MelosysEessiMeldingMapper mapper = mappers.get(sedType);

        if (mapper == null) {
            mapper = defaultMapper;
        }

        return mapper;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapperFactory.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding

import no.nav.melosys.eessi.models.SedType
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class MelosysEessiMeldingMapperFactory(
    @Value("\${rina.institusjon-id}") private val rinaInstitusjonId: String
) {

    private val mappers: Map<SedType, MelosysEessiMeldingMapper> = mapOf(
        SedType.A001 to MelosysEessiMeldingMapperA001(),
        SedType.A002 to MelosysEessiMeldingMapperA002(),
        SedType.A003 to MelosysEessiMeldingMapperA003(),
        SedType.A009 to MelosysEessiMeldingMapperA009(),
        SedType.A010 to MelosysEessiMeldingMapperA010(),
        SedType.A011 to MelosysEessiMeldingMapperA011(),
        SedType.X006 to MelosysEessiMeldingMapperX006(rinaInstitusjonId)
    )

    private val defaultMapper: MelosysEessiMeldingMapper = DefaultMapper()

    fun getMapper(sedType: SedType): MelosysEessiMeldingMapper {
        return mappers[sedType] ?: defaultMapper
    }
}
